anvill/src/Lifters/DataLifter.cpp

```diff
@@ -73,82 +73,128 @@ DataLifter::GetOrDeclareData(const GlobalVarDecl &decl,
-  std::stringstream ss2;
-  ss2 << "var_" << std::hex << decl.address << '_'
-      << TranslateType(*type, dl, true);
+  if (found_by_address) {
 
-  // TODO(akshay): use `dl` to figure out the size in bytes of the type. Make
-  //               a `std::string` and initialize it to that size. Use the
-  //               `MemoryProvider` to query each byte offset from
-  //               `decl.address`, and if the byte is available, then write it
-  //               into the string. Inspect the permissions as you read/write
-  //               bytes. Be careful about crossing permission boundaries; i.e.
-  //               if we cross from writable into non-writable, then we should
-  //               treat that as a failure to get the needed bytes. Log a
-  //               warning/error, and we'll treat the var as an uninitialized
-  //               external. Use that same permissions tracking to set whether
-  //               or not the global var is constant. Try to factor this out
-  //               into one or more functions and not do it all right here.
-  //               Finally, if and when all bytes are read, and no permission
-  //               boundaries are crossed, go and invoke the data lifter.
-  const auto gv = new llvm::GlobalVariable(
-      *options.module, type, false, llvm::GlobalValue::ExternalLinkage,
-      llvm::Constant::getNullValue(type), ss2.str());
+    // We'll use an existing declaration, create a new GEP, and cast as necessary.
+    const auto [base, offset] =
+        remill::StripAndAccumulateConstantOffsets(dl, found_by_address);
+    if (base) {
 
-  ga->setAliasee(gv);
+      // TODO(pag,alessandro): Something related to `remill::BuildPointerToOffset`
+      //                       or `remill::BuildIndexes` to return an initialized
+      //                       global alias.
+      // Key issues:
+      //
+      //    Need to recursively tell the `lifter_context` about all entities
+      //    along the way, kind of like how `FindBaseAndOffset` goes and finds
+      //    them.
+
+      if (offset > 0) {
+
+        // dummy IRBuilder to reuse `remill::BuildPointerToOffset`
+        llvm::IRBuilder<> ir(options.module->getContext());
+        auto gv =
+            llvm::dyn_cast<llvm::GlobalValue>(llvm::dyn_cast<llvm::Constant>(
+                remill::BuildPointerToOffset(ir, base, offset, type)));
+        lifter_context.AddEntity(gv, decl.address);
+
+        ga->setAliasee(gv);
+      } else {
+
+        LOG_IF(ERROR, offset < 0)
+            << "Found negative offset for variable at " << std::hex
+            << decl.address << std::dec
+            << " cast and return to the correct data type.";
+
+        auto gv = llvm::dyn_cast<llvm::GlobalValue>(
+            llvm::ConstantExpr::getBitCast(found_by_address, type));
+        lifter_context.AddEntity(gv, decl.address);
+
+        ga->setAliasee(gv);
+      }
+
+      return ga;
+    }
```

Comment:
I think here you want an `else {` to handle the `!base` case. You can probably do a bitcase there.